### ISSUE TITLE
Fatal failure at proof_node_manager

### ISSUE THREAD

**Issue by rodrigo7491** [cvc5/cvc5#9770]
**Describe the bug**
When running cvc5 in proof-producing mode with the instance in [ex.zip](https://github.com/cvc5/cvc5/files/11554414/ex.zip), the following error is generated:

```
Fatal failure within std::shared_ptr<cvc5::internal::ProofNode> cvc5::internal::ProofNodeManager::mkScope(std::shared_ptr<cvc5::internal::ProofNode>, std::vector<cvc5::internal::NodeTemplate<true> >&, bool, bool, cvc5::internal::Node) at /home/runner/work/cvc5/cvc5/src/proof/proof_node_manager.cpp:229
Unreachable code reached Generated a proof that is not closed by the scope: 
Free assumption: (let ((_let_1 (not (>= F 0)))) (or _let_1 (or _let_1 (not (>= W 1)) (not (>= N 2)))))
- assumption: (and (and (inv_main86 K U B P F M W N L A C G I T O V Z X R Y Q S E J H) (and (= E 0) (= D 1) (not (= Q 0)))) (not (inv_main92 K U B P F M W N L A C G I T O V Z X R Y Q D E J H)))
- assumption: (= inv_main92 (lambda ((|x#25| Int) (|x#26| Int) (|x#27| Int) (|x#28| Int) (|x#29| Int) (|x#30| Int) (|x#31| Int) (|x#32| Int) (|x#33| Int) (|x#34| Int) (|x#35| Int) (|x#36| Int) (|x#37| Int) (|x#38| Int) (|x#39| Int) (|x#40| Int) (|x#41| Int) (|x#42| Int) (|x#43| Int) (|x#44| Int) (|x#45| Int) (|x#46| Int) (|x#47| Int) (|x#48| Int) (|x#49| Int)) (let ((_let_1 (<= (- 1) (* (- 1) |x#48|)))) (let ((_let_2 (<= 0 (* (- 1) |x#47|)))) (let ((_let_3 (<= 1 |x#48|))) (let ((_let_4 (<= 0 (* (- 1) |x#45|)))) (let ((_let_5 (not _let_4))) (let ((_let_6 (<= 0 |x#45|))) (let ((_let_7 (not _let_6))) (let ((_let_8 (<= 0 |x#47|))) (let ((_let_9 (<= (- 1) (* (- 1) |x#46|)))) (let ((_let_10 (<= 1 |x#46|))) (let ((_let_11 (<= (- 1) (* (- 1) |x#44|)))) (let ((_let_12 (<= 0 |x#43|))) (let ((_let_13 (<= 0 (* (- 1) |x#43|)))) (let ((_let_14 (<= 1 |x#44|))) (let ((_let_15 (* (- 1) |x#41|))) (let ((_let_16 (<= 1 _let_15))) (let ((_let_17 (<= 1 |x#41|))) (let ((_let_18 (<= 0 _let_15))) (let ((_let_19 (<= (- 1) (* (- 1) |x#42|)))) (let ((_let_20 (<= 0 |x#41|))) (let ((_let_21 (<= 1 |x#42|))) (let ((_let_22 (<= (- 1) (* (- 1) |x#40|)))) (let ((_let_23 (<= 0 |x#39|))) (let ((_let_24 (<= 0 (* (- 1) |x#39|)))) (let ((_let_25 (* (- 1) |x#37|))) (let ((_let_26 (<= 1 _let_25))) (let ((_let_27 (<= 1 |x#37|))) (let ((_let_28 (<= 1 |x#40|))) (let ((_let_29 (<= 0 _let_25))) (let ((_let_30 (<= (- 1) (* (- 1) |x#38|)))) (let ((_let_31 (<= 0 |x#37|))) (let ((_let_32 (<= 1 |x#38|))) (let ((_let_33 (<= (- 1) (* (- 1) |x#36|)))) (let ((_let_34 (<= 0 |x#35|))) (let ((_let_35 (<= 0 (* (- 1) |x#35|)))) (let ((_let_36 (<= 1 |x#36|))) (let ((_let_37 (* (- 1) |x#33|))) (let ((_let_38 (<= 1 _let_37))) (let ((_let_39 (<= 1 |x#33|))) (let ((_let_40 (<= 0 _let_37))) (let ((_let_41 (<= (- 1) (* (- 1) |x#34|)))) (let ((_let_42 (<= 0 |x#33|))) (let ((_let_43 (<= 1 |x#34|))) (let ((_let_44 (<= (- 1) (* (- 1) |x#32|)))) (let ((_let_45 (<= 0 |x#31|))) (let ((_let_46 (<= 0 (* (- 1) |x#31|)))) (let ((_let_47 (* (- 1) |x#29|))) (let ((_let_48 (<= 1 _let_47))) (let ((_let_49 (<= 1 |x#29|))) (let ((_let_50 (<= 1 |x#32|))) (let ((_let_51 (<= 0 _let_47))) (let ((_let_52 (<= (- 1) (* (- 1) |x#30|)))) (let ((_let_53 (<= 0 |x#29|))) (let ((_let_54 (<= 1 |x#30|))) (let ((_let_55 (<= 0 (* (- 1) |x#27|)))) (let ((_let_56 (<= 1 |x#28|))) (let ((_let_57 (<= 0 |x#27|))) (let ((_let_58 (<= 1 |x#25|))) (let ((_let_59 (<= 0 (* (- 1) |x#25|)))) (let ((_let_60 (<= 0 |x#25|))) (let ((_let_61 (<= 1 |x#26|))) (let ((_let_62 (<= (- 1) (* (- 1) |x#26|)))) (let ((_let_63 (and _let_62 (or _let_61 (and _let_60 _let_59))))) (and (or _let_63 (and _let_61 (or _let_62 _let_63)) (and _let_60 (or _let_59 _let_63))) (or _let_58 _let_57 _let_56) (or _let_58 _let_56 _let_55) (<= (- 1) (* (- 1) |x#28|)) (or _let_57 _let_56) (or _let_56 _let_55) (or _let_53 _let_54) (or _let_54 _let_51) (or _let_53 _let_52) (or _let_51 _let_52) (or _let_48 _let_49 _let_45 _let_50) (or _let_45 _let_50) (or _let_48 _let_49 _let_50 _let_46) (or _let_51 _let_45 _let_44) (or _let_48 _let_45 _let_44) (or _let_50 _let_46) (or _let_48 _let_49 _let_46 _let_44) (or _let_48 _let_46 _let_44) (or _let_46 _let_44) (or _let_45 _let_44) (or _let_42 _let_43) (or _let_43 _let_40) (or _let_42 _let_41) (or _let_40 _let_41) (or _let_38 _let_39 _let_34 _let_36) (or _let_34 _let_36) (or _let_38 _let_39 _let_36 _let_35) (or _let_40 _let_34 _let_33) (or _let_38 _let_34 _let_33) (or _let_38 _let_36 _let_35) (or _let_38 _let_39 _let_35 _let_33) (or _let_38 _let_35 _let_33) (or _let_36 _let_35) (or _let_35 _let_33) (or _let_34 _let_33) (or _let_31 _let_32) (or _let_32 _let_29) (or _let_31 _let_30) (or _let_29 _let_30) (or _let_26 _let_27 _let_23 _let_28) (or _let_23 _let_28) (or _let_26 _let_27 _let_28 _let_24) (or _let_29 _let_23 _let_22) (or _let_26 _let_23 _let_22) (or _let_28 _let_24) (or _let_26 _let_27 _let_24 _let_22) (or _let_26 _let_24 _let_22) (or _let_24 _let_22) (or _let_23 _let_22) (or _let_20 _let_21) (or _let_21 _let_18) (or _let_20 _let_19) (or _let_18 _let_19) (or _let_16 _let_17 _let_12 _let_14) (or _let_12 _let_14) (or _let_16 _let_17 _let_14 _let_13) (or _let_18 _let_12 _let_11) (or _let_16 _let_12 _let_11) (or _let_16 _let_14 _let_13) (or _let_16 _let_17 _let_13 _let_11) (or _let_16 _let_13 _let_11) (or _let_14 _let_13) (or _let_13 _let_11) (or _let_12 _let_11) (or _let_10 _let_6) (or _let_10 _let_4) (or _let_6 _let_9) (or _let_4 _let_9) (or _let_7 _let_5 _let_3 _let_8) (or _let_3 _let_8) (or _let_7 _let_5 _let_3 _let_2) (or _let_8 _let_1) (or _let_7 _let_5 _let_2 _let_1) (or _let_3 _let_2) (or _let_2 _let_1)))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))
- assumption: (= inv_main86 (lambda ((|x#350| Int) (|x#351| Int) (|x#352| Int) (|x#353| Int) (|x#354| Int) (|x#355| Int) (|x#356| Int) (|x#357| Int) (|x#358| Int) (|x#359| Int) (|x#360| Int) (|x#361| Int) (|x#362| Int) (|x#363| Int) (|x#364| Int) (|x#365| Int) (|x#366| Int) (|x#367| Int) (|x#368| Int) (|x#369| Int) (|x#370| Int) (|x#371| Int) (|x#372| Int) (|x#373| Int) (|x#374| Int)) (let ((_let_1 (<= (- 1) (* (- 1) |x#369|)))) (let ((_let_2 (<= 0 |x#368|))) (let ((_let_3 (<= 0 |x#366|))) (let ((_let_4 (<= 0 (* (- 1) |x#368|)))) (let ((_let_5 (<= 1 |x#369|))) (let ((_let_6 (<= 0 (* (- 1) |x#366|)))) (let ((_let_7 (not _let_6))) (let ((_let_8 (not _let_3))) (let ((_let_9 (<= (- 1) (* (- 1) |x#367|)))) (let ((_let_10 (<= 1 |x#367|))) (let ((_let_11 (<= (- 1) (* (- 1) |x#365|)))) (let ((_let_12 (<= 0 |x#364|))) (let ((_let_13 (<= 0 (* (- 1) |x#364|)))) (let ((_let_14 (* (- 1) |x#362|))) (let ((_let_15 (<= 1 _let_14))) (let ((_let_16 (<= 1 |x#362|))) (let ((_let_17 (<= 1 |x#365|))) (let ((_let_18 (<= 0 _let_14))) (let ((_let_19 (<= (- 1) (* (- 1) |x#363|)))) (let ((_let_20 (<= 0 |x#362|))) (let ((_let_21 (<= 1 |x#363|))) (let ((_let_22 (<= (- 1) (* (- 1) |x#361|)))) (let ((_let_23 (<= 0 |x#360|))) (let ((_let_24 (<= 0 (* (- 1) |x#360|)))) (let ((_let_25 (<= 1 |x#361|))) (let ((_let_26 (* (- 1) |x#358|))) (let ((_let_27 (<= 1 _let_26))) (let ((_let_28 (<= 1 |x#358|))) (let ((_let_29 (<= 0 _let_26))) (let ((_let_30 (<= (- 1) (* (- 1) |x#359|)))) (let ((_let_31 (<= 0 |x#358|))) (let ((_let_32 (<= 1 |x#359|))) (let ((_let_33 (<= (- 1) (* (- 1) |x#357|)))) (let ((_let_34 (<= 0 |x#356|))) (let ((_let_35 (<= 0 (* (- 1) |x#356|)))) (let ((_let_36 (* (- 1) |x#354|))) (let ((_let_37 (<= 1 _let_36))) (let ((_let_38 (<= 1 |x#354|))) (let ((_let_39 (<= 1 |x#357|))) (let ((_let_40 (<= 0 _let_36))) (let ((_let_41 (<= (- 1) (* (- 1) |x#355|)))) (let ((_let_42 (<= 0 |x#354|))) (let ((_let_43 (<= 1 |x#355|))) (let ((_let_44 (<= 0 (* (- 1) |x#352|)))) (let ((_let_45 (<= 1 |x#353|))) (let ((_let_46 (<= 0 |x#352|))) (let ((_let_47 (<= 1 |x#350|))) (let ((_let_48 (* (- 1) |x#350|))) (let ((_let_49 (<= 1 _let_48))) (let ((_let_50 (<= 1 |x#351|))) (and (or (<= 0 |x#350|) _let_50) (or _let_50 (<= 0 _let_48)) (<= (- 1) (* (- 1) |x#351|)) (or _let_49 _let_47 _let_46 _let_45) (or _let_47 _let_46 _let_45) (or _let_49 _let_47 _let_45 _let_44) (or _let_47 _let_45 _let_44) (<= (- 1) (* (- 1) |x#353|)) (or _let_46 _let_45) (or _let_45 _let_44) (or _let_42 _let_43) (or _let_43 _let_40) (or _let_42 _let_41) (or _let_40 _let_41) (or _let_37 _let_38 _let_34 _let_39) (or _let_34 _let_39) (or _let_37 _let_38 _let_39 _let_35) (or _let_40 _let_34 _let_33) (or _let_37 _let_34 _let_33) (or _let_39 _let_35) (or _let_37 _let_38 _let_35 _let_33) (or _let_37 _let_35 _let_33) (or _let_35 _let_33) (or _let_34 _let_33) (or _let_31 _let_32) (or _let_32 _let_29) (or _let_31 _let_30) (or _let_29 _let_30) (or _let_27 _let_28 _let_23 _let_25) (or _let_23 _let_25) (or _let_27 _let_28 _let_25 _let_24) (or _let_29 _let_23 _let_22) (or _let_27 _let_23 _let_22) (or _let_27 _let_25 _let_24) (or _let_27 _let_28 _let_24 _let_22) (or _let_27 _let_24 _let_22) (or _let_25 _let_24) (or _let_24 _let_22) (or _let_23 _let_22) (or _let_20 _let_21) (or _let_21 _let_18) (or _let_20 _let_19) (or _let_18 _let_19) (or _let_15 _let_16 _let_12 _let_17) (or _let_12 _let_17) (or _let_15 _let_16 _let_17 _let_13) (or _let_18 _let_12 _let_11) (or _let_15 _let_12 _let_11) (or _let_17 _let_13) (or _let_15 _let_16 _let_13 _let_11) (or _let_15 _let_13 _let_11) (or _let_13 _let_11) (or _let_12 _let_11) (or _let_3 _let_10) (or _let_10 _let_6) (or _let_3 _let_9) (or _let_6 _let_9) (or _let_8 _let_7 _let_2 _let_5) (or _let_2 _let_5) (or _let_8 _let_7 _let_5 _let_4) (or _let_6 _let_2 _let_1) (or _let_8 _let_7 _let_2 _let_1) (or _let_6 _let_5 _let_4) (or _let_8 _let_7 _let_4 _let_1) (or _let_6 _let_4 _let_1) (or _let_3 _let_5 _let_4) (or _let_3 _let_4 _let_1) (or _let_3 _let_2 _let_1))))))))))))))))))))))))))))))))))))))))))))))))))))))
Use -t dump-proof-error for details on proof


Aborted (core dumped)
```

The error happens with both `--proof-format=alethe` and `--proof-format=lfsc`, and does not occur if `(get-proof)` is commented out. Using `-t dump-proof-error` led to `(error "Error in option parsing: trace tags not available in non-tracing builds")`; I am using a release binary. Minimization with `ddsmt ex.smt2 ex_ddsmt.smt2 cvc5` does not seem to have had an effect.

**Command line arguments**: `--produce-proofs --proof-format=alethe` and `--produce-proofs --proof-format=lfsc`
**cvc5 version/commit**: v1.0.5
**Operating system**: Linux Mint 20.2 and Ubuntu 20.04.1

**Comment by ajreynol** [comment]
Thanks for the report. 

I was able get ddsmt to minimize the file:
```
(set-option :produce-proofs true)
(set-logic QF_LIA)
(declare-fun K () Int)
(declare-fun P () Int)
(declare-fun R () Int)
(declare-fun X () Int)
(declare-fun Z () Int)
(assert (and (or (not (or (<= 1 P) (<= 0 K) (<= 1 R))) (and (distinct 0 Z) (= 0 Z)) (= 0 X)) (or (<= 1 P) (<= 0 K) (<= 1 R)) (or (= 1 Z) (< 0 X)) (or (< 0 X) (= 0 Z))))
(check-sat)
(get-proof)
```
I will look into this shortly.

**Comment by ajreynol** [comment]
@HanielB The free assumption appears to be coming from the SAT proof.  Try `--check-proofs --proof-check=eager`.  I get backtrace: 
```
#3  0x000000000088cf84 in cvc5::internal::FatalStream::~FatalStream (this=this@entry=0x7fffffffc4af, __in_chrg=<optimized out>) at /space/ajreynol/cvc5-ajr/src/base/check.cpp:32
#4  0x00000000008dd407 in cvc5::internal::ensureClosedWrtInternal (opts=..., proven=..., pg=pg@entry=0x0, pnp=pnp@entry=0x22ed370, assumps=..., c=c@entry=0x178dc92 "sat-proof", ctx=<optimized out>, reqGen=false) at /space/ajreynol/cvc5-ajr/src/proof/proof_ensure_closed.cpp:141
#5  0x00000000008ddb08 in cvc5::internal::pfnEnsureClosedWrt (opts=..., pn=pn@entry=0x22ed370, assumps=..., c=c@entry=0x178dc92 "sat-proof", ctx=ctx@entry=0x1795f77 "PropPfManager::checkProof") at /space/ajreynol/cvc5-ajr/src/proof/proof_ensure_closed.cpp:187
#6  0x00000000008fb2ea in cvc5::internal::prop::PropPfManager::checkProof (this=0x20a69d0, assertions=...) at /space/ajreynol/cvc5-ajr/src/prop/prop_proof_manager.cpp:66
#7  0x00000000008f8cf3 in cvc5::internal::prop::PropEngine::checkProof (this=this@entry=0x20653a0, assertions=...) at /space/ajreynol/cvc5-ajr/src/prop/prop_engine.cpp:692
#8  0x00000000004fccad in cvc5::internal::SolverEngine::checkProof (this=this@entry=0x1ef7920) at /space/ajreynol/cvc5-ajr/src/smt/solver_engine.cpp:1316
#9  0x0000000000506783 in cvc5::internal::SolverEngine::checkSatInternal (this=this@entry=0x1ef7920, assumptions=...) at /space/ajreynol/cvc5-ajr/src/smt/solver_engine.cpp:757
#10 0x00000000005068fb in cvc5::internal::SolverEngine::checkSat (this=0x1ef7920) at /space/ajreynol/cvc5-ajr/src/smt/solver_engine.cpp:700
#11 0x000000000049b5f7 in cvc5::Solver::checkSat (this=this@entry=0x1ef0da0) at /space/ajreynol/cvc5-ajr/src/api/cpp/cvc5.cpp:6635
#12 0x00000000011bcdc0 in cvc5::parser::CheckSatCommand::invoke (this=0x2288870, solver=0x1ef0da0, sm=<optimized out>) at /space/ajreynol/cvc5-ajr/src/parser/api/cpp/command.cpp:399
#13 0x00000000011b8151 in cvc5::parser::Command::invoke (this=0x2288870, solver=0x1ef0da0, sm=<optimized out>, out=...) at /space/ajreynol/cvc5-ajr/src/parser/api/cpp/command.cpp:164
#14 0x0000000000435add in cvc5::main::CommandExecutor::solverInvoke (this=this@entry=0x1f04430, solver=0x1ef0da0, sm=sm@entry=0x1f0b900, cmd=cmd@entry=0x2288870, out=...) at /space/ajreynol/cvc5-ajr/src/main/command_executor.cpp:211
#15 0x00000000004362bf in cvc5::main::CommandExecutor::doCommandSingleton (this=0x1f04430, cmd=0x2288870) at /space/ajreynol/cvc5-ajr/src/main/command_executor.cpp:109
#16 0x0000000000435c53 in cvc5::main::CommandExecutor::doCommand (this=this@entry=0x1f04430, cmd=0x2288870) at /space/ajreynol/cvc5-ajr/src/main/command_executor.cpp:98
#17 0x0000000000442991 in cvc5::main::CommandExecutor::doCommand (this=0x1f04430, cmd=...) at /space/ajreynol/cvc5-ajr/src/./main/command_executor.h:73
#18 0x000000000043a8cc in cvc5::main::ExecutionContext::solveContinuous (this=this@entry=0x7fffffffd080, parser=0x1f18750, stopAtSetLogic=stopAtSetLogic@entry=false) at /space/ajreynol/cvc5-ajr/src/main/portfolio_driver.cpp:61
#19 0x00000000004413e0 in cvc5::main::PortfolioDriver::solve (this=this@entry=0x7fffffffd140, executor=...) at /space/ajreynol/cvc5-ajr/src/main/portfolio_driver.cpp:422
#20 0x0000000000433b96 in runCvc5 (argc=argc@entry=13, argv=argv@entry=0x7fffffffd598, solver=...) at /space/ajreynol/cvc5-ajr/src/main/driver_unified.cpp:236
#21 0x000000000043544c in main (argc=13, argv=0x7fffffffd598) at /space/ajreynol/cvc5-ajr/src/main/main.cpp:36
```
On the minimized example above, the relevant subproof in the final proof appears to be:
```
(MACRO_RESOLUTION_TRUST (ASSUME :args ((or (>= P 1) (or (>= P 1) (>= K 0) (>= R 1))))) (CNF_OR_NEG :args ((or (>= P 1) (>= K 0) (>= R 1)) 0)) :args ((or (>= P 1) (>= K 0) (>= R 1)) true (>= P 1)))
```

**Comment by someplaceguy** [comment]
FYI: I've filed issue #10279 which is also leading to the same fatal failure, but on the surface it appears to be due to a different issue.

That said, I know nothing about how cvc5 is implemented, so maybe the root cause really is the same...

**Comment by HanielB** [comment]
This issue was fixed in #10683, which also already added regressions.
